src/collections/componentSetBuilder.ts

```diff
@@ -0,0 +1,171 @@
+        logger.debug(`Building ComponentSet from sourcepath: ${sourcepath.join(', ')}`);
+        const fsPaths: string[] = sourcepath.map((filepath) => {
+          if (!fs.existsSync(filepath)) {
+            throw new SfdxError(`The sourcepath "${filepath}" is not a valid source file path.`);
+          }
+          return path.resolve(filepath);
+        });
+        componentSet = ComponentSet.fromSource({ fsPaths });
+      }
+
+      // Return empty ComponentSet and use packageNames in the connection via `.retrieve` options
+      if (packagenames) {
+        logger.debug(`Building ComponentSet for packagenames: ${packagenames.toString()}`);
+        componentSet ??= new ComponentSet();
+      }
+
+      // Resolve manifest with source in package directories.
+      if (manifest) {
+        logger.debug(`Building ComponentSet from manifest: ${manifest.manifestPath}`);
+        const directoryPaths = options.manifest.directoryPaths;
+        logger.debug(`Searching in packageDir: ${directoryPaths.join(', ')} for matching metadata`);
+        componentSet = await ComponentSet.fromManifest({
+          manifestPath: manifest.manifestPath,
+          resolveSourcePaths: directoryPaths,
+          forceAddWildcards: true,
+          destructivePre: options.manifest.destructiveChangesPre,
+          destructivePost: options.manifest.destructiveChangesPost,
+        });
+      }
+
+      // Resolve metadata entries with source in package directories.
+      if (metadata) {
+        logger.debug(`Building ComponentSet from metadata: ${metadata.metadataEntries.toString()}`);
+        const registry = new RegistryAccess();
+        const compSetFilter = new ComponentSet();
+        componentSet ??= new ComponentSet();
+
+        // Build a Set of metadata entries
+        metadata.metadataEntries.forEach((rawEntry) => {
+          const splitEntry = rawEntry.split(':').map((entry) => entry.trim());
+          // The registry will throw if it doesn't know what this type is.
+          registry.getTypeByName(splitEntry[0]);
+          const entry = {
+            type: splitEntry[0],
+            fullName: splitEntry.length === 1 ? '*' : splitEntry[1],
+          };
+          // Add to the filtered ComponentSet for resolved source paths,
+          // and the unfiltered ComponentSet to build the correct manifest.
+          compSetFilter.add(entry);
+          componentSet.add(entry);
+        });
+
+        const directoryPaths = options.metadata.directoryPaths;
+        logger.debug(`Searching for matching metadata in directories: ${directoryPaths.join(', ')}`);
+        const resolvedComponents = ComponentSet.fromSource({ fsPaths: directoryPaths, include: compSetFilter });
+        componentSet.forceIgnoredPaths = resolvedComponents.forceIgnoredPaths;
+        for (const comp of resolvedComponents) {
+          componentSet.add(comp);
+        }
+      }
+
+      // Resolve metadata entries with an org connection
+      if (org) {
```

Comment:
This is where I think we'll have to add some validation of `ComponentSetOptions`.  I can't think of a reason why someone would want to combine the `org` option with others, but line 121 below will clobber `componentSet` anyway.  Either the reassignment should change to be additive or we need to validate how people are combining `ComponentSetOptions`.